crates/shadowsocks/src/relay/udprelay/proxy_socket.rs

```diff
@@ -254,6 +251,76 @@ impl ProxySocket {
         Ok(send_len)
     }
 
+    pub fn poll_send(&self, addr: &Address, payload: &[u8], cx: &mut Context<'_>) -> Poll<ProxySocketResult<usize>> {
```

Comment:
Since you have moved the main logic from `send` to `poll_send`, then you should use `futures::future::poll_fn` to reimplement `send` with `poll_send`.

Reply:
i think send and poll send are essentially different family function calls which should be calling each other internally.

as inside the send call, timeout is taken care of while the poll_send doesn't have that logic and just return different poll state.

wdyt?

Reply:
tokio also seems to implmement these two separately https://docs.rs/tokio/latest/src/tokio/net/udp.rs.html#768-773

Reply:
> as inside the send call, timeout is taken care of

Hmm, I don't think so. Async calls doesn't have timeouts.

Reply:
https://github.com/tokio-rs/tokio/blob/a66884a2fb80d1180451706f3c3e006a3fdcb036/tokio/src/net/tcp/split.rs#L137-L140

I am not really sure why tokio implements UdpSocket's `recv` separately. But some part of its code actually uses `poll_fn` to reuse `poll_*`s.

Reply:
right i was more referring to this implementation detail:

https://github.com/shadowsocks/shadowsocks-rust/blob/ab6a6f6178d3a29e9eea16fd73b21c80a7a77278/crates/shadowsocks/src/relay/udprelay/proxy_socket.rs#L340-L347

it would a bit tricky to check timeout on the poll_fn, I suppose?

Reply:
i think if it's just a simple wrap it might be fine, otherwise maybe implement different logic separately

Reply:
Ah you are right, I just forgot there is a `self.recv_timeout`. Well yes, it will be very tricky to implement timeout in `poll_recv`. You should mention that in the doc.

Reply:
or just simply 

```rust
let recv_n = match self.recv_timeout { 
     None => poll_fn(...).await?, 
     Some(d) => match time::timeout(d, poll_fn(...).await? { 
         Ok(Ok(l)) => l, 
         Ok(Err(err)) => return Err(err.into()), 
         Err(..) => return Err(io::Error::from(ErrorKind::TimedOut).into()), 
     }, 
 }; 
```

i'm happy to go with either way

Reply:
I am ok with your simple solution.